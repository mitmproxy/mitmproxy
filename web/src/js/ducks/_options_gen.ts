/** Auto-generated by test_app.py:test_generate_options_js */
export interface OptionsState {
    add_upstream_certs_to_client_chain: boolean
    allow_hosts: string[]
    anticache: boolean
    anticomp: boolean
    block_global: boolean
    block_list: string[]
    block_private: boolean
    body_size_limit: string | undefined
    cert_passphrase: string | undefined
    certs: string[]
    ciphers_client: string | undefined
    ciphers_server: string | undefined
    client_certs: string | undefined
    client_replay: string[]
    client_replay_concurrency: number
    command_history: boolean
    confdir: string
    connect_addr: string | undefined
    connection_strategy: string
    console_focus_follow: boolean
    content_view_lines_cutoff: number
    dns_listen_host: string
    dns_listen_port: number
    dns_mode: string
    dns_server: boolean
    export_preserve_original_ip: boolean
    http2: boolean
    http2_ping_keepalive: number
    ignore_hosts: string[]
    intercept: string | undefined
    intercept_active: boolean
    keep_host_header: boolean
    key_size: number
    listen_host: string
    listen_port: number
    map_local: string[]
    map_remote: string[]
    mode: string
    modify_body: string[]
    modify_headers: string[]
    normalize_outbound_headers: boolean
    onboarding: boolean
    onboarding_host: string
    onboarding_port: number
    proxy_debug: boolean
    proxyauth: string | undefined
    rawtcp: boolean
    readfile_filter: string | undefined
    rfile: string | undefined
    save_stream_file: string | undefined
    save_stream_filter: string | undefined
    scripts: string[]
    server: boolean
    server_replay: string[]
    server_replay_ignore_content: boolean
    server_replay_ignore_host: boolean
    server_replay_ignore_params: string[]
    server_replay_ignore_payload_params: string[]
    server_replay_ignore_port: boolean
    server_replay_kill_extra: boolean
    server_replay_nopop: boolean
    server_replay_refresh: boolean
    server_replay_use_headers: string[]
    showhost: boolean
    ssl_insecure: boolean
    ssl_verify_upstream_trusted_ca: string | undefined
    ssl_verify_upstream_trusted_confdir: string | undefined
    stickyauth: string | undefined
    stickycookie: string | undefined
    stream_large_bodies: string | undefined
    tcp_hosts: string[]
    termlog_verbosity: string
    tls_version_client_max: string
    tls_version_client_min: string
    tls_version_server_max: string
    tls_version_server_min: string
    upstream_auth: string | undefined
    upstream_cert: boolean
    validate_inbound_headers: boolean
    view_filter: string | undefined
    view_order: string
    view_order_reversed: boolean
    web_columns: string[]
    web_debug: boolean
    web_host: string
    web_open_browser: boolean
    web_port: number
    web_static_viewer: string | undefined
    websocket: boolean
}

export type Option = keyof OptionsState

export const defaultState: OptionsState = {
    add_upstream_certs_to_client_chain: false,
    allow_hosts: [],
    anticache: false,
    anticomp: false,
    block_global: true,
    block_list: [],
    block_private: false,
    body_size_limit: undefined,
    cert_passphrase: undefined,
    certs: [],
    ciphers_client: undefined,
    ciphers_server: undefined,
    client_certs: undefined,
    client_replay: [],
    client_replay_concurrency: 1,
    command_history: true,
    confdir: "~/.mitmproxy",
    connect_addr: undefined,
    connection_strategy: "eager",
    console_focus_follow: false,
    content_view_lines_cutoff: 512,
    dns_listen_host: "",
    dns_listen_port: 53,
    dns_mode: "regular",
    dns_server: false,
    export_preserve_original_ip: false,
    http2: true,
    http2_ping_keepalive: 58,
    ignore_hosts: [],
    intercept: undefined,
    intercept_active: false,
    keep_host_header: false,
    key_size: 2048,
    listen_host: "",
    listen_port: 8080,
    map_local: [],
    map_remote: [],
    mode: "regular",
    modify_body: [],
    modify_headers: [],
    normalize_outbound_headers: true,
    onboarding: true,
    onboarding_host: "mitm.it",
    onboarding_port: 80,
    proxy_debug: false,
    proxyauth: undefined,
    rawtcp: true,
    readfile_filter: undefined,
    rfile: undefined,
    save_stream_file: undefined,
    save_stream_filter: undefined,
    scripts: [],
    server: true,
    server_replay: [],
    server_replay_ignore_content: false,
    server_replay_ignore_host: false,
    server_replay_ignore_params: [],
    server_replay_ignore_payload_params: [],
    server_replay_ignore_port: false,
    server_replay_kill_extra: false,
    server_replay_nopop: false,
    server_replay_refresh: true,
    server_replay_use_headers: [],
    showhost: false,
    ssl_insecure: false,
    ssl_verify_upstream_trusted_ca: undefined,
    ssl_verify_upstream_trusted_confdir: undefined,
    stickyauth: undefined,
    stickycookie: undefined,
    stream_large_bodies: undefined,
    tcp_hosts: [],
    termlog_verbosity: "info",
    tls_version_client_max: "UNBOUNDED",
    tls_version_client_min: "TLS1_2",
    tls_version_server_max: "UNBOUNDED",
    tls_version_server_min: "TLS1_2",
    upstream_auth: undefined,
    upstream_cert: true,
    validate_inbound_headers: true,
    view_filter: undefined,
    view_order: "time",
    view_order_reversed: false,
    web_columns: ["tls", "icon", "path", "method", "status", "size", "time"],
    web_debug: false,
    web_host: "127.0.0.1",
    web_open_browser: true,
    web_port: 8081,
    web_static_viewer: "",
    websocket: true,
}
