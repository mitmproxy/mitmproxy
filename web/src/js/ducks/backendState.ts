/**
 * serverState houses properties about the current mitmproxy instance that are not options,
 * e.g. the list of available content views or the current version.
 */

export const RECEIVE = 'STATE_RECEIVE'
export const UPDATE = 'STATE_UPDATE'

export interface ServerInfo {
    description: string
    full_spec: string
    is_running: boolean
    last_exception: string | null
    listen_addrs: [string, number][]
    type: string
    wireguard_conf?: string
}

export interface BackendState {
    available: boolean
    version: string
    contentViews: string[]
    servers: ServerInfo[]
}

export const defaultState: BackendState = {
    available: false,
    version: "",
    contentViews: [],
    servers: [],
};

export default function reducer(state = defaultState, action): BackendState {
    switch (action.type) {
        case RECEIVE:
        case UPDATE:
            return {
                ...state,
                available: true,
                ...action.data
            }
        default:
            return state
    }
}
