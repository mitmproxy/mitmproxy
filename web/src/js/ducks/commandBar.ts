export const TOGGLE_VISIBILITY = "COMMANDBAR_TOGGLE_VISIBILITY";

interface CommandBarState {
    visible: boolean;
}

export const defaultState: CommandBarState = {
    visible: false,
};

export default function reducer(state = defaultState, action): CommandBarState {
    switch (action.type) {
        case TOGGLE_VISIBILITY:
            return {
                ...state,
                visible: !state.visible,
            };

        default:
            return state;
    }
}

export function toggleVisibility() {
    return { type: TOGGLE_VISIBILITY };
}
