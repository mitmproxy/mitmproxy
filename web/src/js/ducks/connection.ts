export enum ConnectionState {
    INIT = "CONNECTION_INIT",
    FETCHING = "CONNECTION_FETCHING", // WebSocket is established, but still fetching resources.
    ESTABLISHED = "CONNECTION_ESTABLISHED",
    ERROR = "CONNECTION_ERROR",
    OFFLINE = "CONNECTION_OFFLINE", // indicates that there is no live (websocket) backend.
}

interface ConnState {
    state: ConnectionState;
    message?: string;
}

const defaultState: ConnState = {
    state: ConnectionState.INIT,
    message: undefined,
};

export default function reducer(state = defaultState, action): ConnState {
    switch (action.type) {
        case ConnectionState.ESTABLISHED:
        case ConnectionState.FETCHING:
        case ConnectionState.ERROR:
        case ConnectionState.OFFLINE:
            return {
                state: action.type,
                message: action.message,
            };

        default:
            return state;
    }
}

export function startFetching() {
    return { type: ConnectionState.FETCHING };
}

export function connectionEstablished() {
    return { type: ConnectionState.ESTABLISHED };
}

export function connectionError(message) {
    return { type: ConnectionState.ERROR, message };
}

export function setOffline() {
    return { type: ConnectionState.OFFLINE };
}
