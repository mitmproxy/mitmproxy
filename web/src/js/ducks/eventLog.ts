import * as store from "./utils/store";

export const ADD = "EVENTS_ADD";
export const RECEIVE = "EVENTS_RECEIVE";
export const TOGGLE_VISIBILITY = "EVENTS_TOGGLE_VISIBILITY";
export const TOGGLE_FILTER = "EVENTS_TOGGLE_FILTER";

type LogLevel = "debug" | "info" | "web" | "warn" | "error";

export interface EventLogItem extends store.Item {
    message: string;
    level: LogLevel;
}

interface EventLogState extends store.State<EventLogItem> {
    visible: boolean;
    filters: { [level in LogLevel]: boolean };
}

const defaultState: EventLogState = {
    visible: false,
    filters: { debug: false, info: true, web: true, warn: true, error: true },
    ...store.defaultState,
};

export default function reduce(
    state: EventLogState = defaultState,
    action
): EventLogState {
    switch (action.type) {
        case TOGGLE_VISIBILITY:
            return {
                ...state,
                visible: !state.visible,
            };

        case TOGGLE_FILTER:
            const filters = {
                ...state.filters,
                [action.filter]: !state.filters[action.filter],
            };
            return {
                ...state,
                filters,
                ...store.reduce(
                    state,
                    store.setFilter<EventLogItem>((log) => filters[log.level])
                ),
            };

        case ADD:
        case RECEIVE:
            return {
                ...state,
                ...store.reduce(
                    state,
                    store[action.cmd](
                        action.data,
                        (log) => state.filters[log.level]
                    )
                ),
            };

        default:
            return state;
    }
}

export function toggleFilter(filter: LogLevel) {
    return { type: TOGGLE_FILTER, filter };
}

export function toggleVisibility() {
    return { type: TOGGLE_VISIBILITY };
}

export function add(message: string, level: LogLevel = "web") {
    let data = {
        id: Math.random().toString(),
        message,
        level,
    };
    return {
        type: ADD,
        cmd: "add",
        data,
    };
}
