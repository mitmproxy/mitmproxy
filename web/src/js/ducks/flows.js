import { fetchApi } from "../utils"
import reduceStore from "./utils/store"
import * as storeActions from "./utils/store"
import Filt from "../filt/filt"
import { RequestUtils } from "../flow/utils"

export const ADD            = 'FLOWS_ADD'
export const UPDATE         = 'FLOWS_UPDATE'
export const REMOVE         = 'FLOWS_REMOVE'
export const RECEIVE        = 'FLOWS_RECEIVE'
export const SELECT         = 'FLOWS_SELECT'
export const SET_FILTER     = 'FLOWS_SET_FILTER'
export const SET_SORT       = 'FLOWS_SET_SORT'
export const SET_HIGHLIGHT  = 'FLOWS_SET_HIGHLIGHT'
export const REQUEST_ACTION = 'FLOWS_REQUEST_ACTION'


const defaultState = {
    highlight: null,
    filter: null,
    sort: { column: null, desc: false },
    selected: [],
    ...reduceStore(undefined, {})
}

export default function reduce(state = defaultState, action) {
    switch (action.type) {

        case ADD:
        case UPDATE:
        case REMOVE:
        case RECEIVE:
            let storeAction = storeActions[action.cmd](
                action.data,
                makeFilter(state.filter),
                makeSort(state.sort)
            )

            let selected = state.selected
            if(action.type === REMOVE && state.selected.includes(action.data)) {
                if(state.selected.length > 1){
                    selected = selected.filter(x => x !== action.data)
                } else {
                    selected = []
                    if (action.data in state.viewIndex && state.view.length > 1) {
                        let currentIndex = state.viewIndex[action.data],
                            nextSelection
                        if(currentIndex === state.view.length -1){ // last row
                            nextSelection = state.view[currentIndex - 1]
                        } else {
                            nextSelection = state.view[currentIndex + 1]
                        }
                        selected.push(nextSelection.id)
                    }
                }
            }

            return {
                ...state,
                selected,
                ...reduceStore(state, storeAction)
            }

        case SET_FILTER:
            return {
                ...state,
                filter: action.filter,
                ...reduceStore(state, storeActions.setFilter(makeFilter(action.filter), makeSort(state.sort)))
            }

        case SET_HIGHLIGHT:
            return {
                ...state,
                highlight: action.highlight
            }

        case SET_SORT:
            return {
                ...state,
                sort: action.sort,
                ...reduceStore(state, storeActions.setSort(makeSort(action.sort)))
            }

        case SELECT:
            return {
                ...state,
                selected: action.flowIds
            }

        default:
            return state
    }
}


const sortKeyFuns = {

    TLSColumn: flow => flow.request.scheme,

    PathColumn: flow => RequestUtils.pretty_url(flow.request),

    MethodColumn: flow => flow.request.method,

    StatusColumn: flow => flow.response && flow.response.status_code,

    TimeColumn: flow => flow.response && flow.response.timestamp_end - flow.request.timestamp_start,

    SizeColumn: flow => {
        let total = flow.request.contentLength
        if (flow.response) {
            total += flow.response.contentLength || 0
        }
        return total
    },
}

export function makeFilter(filter) {
    if (!filter) {
        return
    }
    return Filt.parse(filter)
}

export function makeSort({ column, desc }) {
    const sortKeyFun = sortKeyFuns[column]
    if (!sortKeyFun) {
        return
    }
    return (a, b) => {
        const ka = sortKeyFun(a)
        const kb = sortKeyFun(b)
        if (ka > kb) {
            return desc ? -1 : 1
        }
        if (ka < kb) {
            return desc ? 1 : -1
        }
        return 0
    }
}

export function setFilter(filter) {
    return { type: SET_FILTER, filter }
}

export function setHighlight(highlight) {
    return { type: SET_HIGHLIGHT, highlight }
}

export function setSort(column, desc) {
    return { type: SET_SORT, sort: { column, desc } }
}

export function selectRelative(flows, shift) {
    let currentSelectionIndex = flows.viewIndex[flows.selected[0]]
    let minIndex = 0
    let maxIndex = flows.view.length - 1
    let newIndex
    if (currentSelectionIndex === undefined) {
        newIndex = (shift < 0) ? minIndex : maxIndex
    } else {
        newIndex = currentSelectionIndex + shift
        newIndex = window.Math.max(newIndex, minIndex)
        newIndex = window.Math.min(newIndex, maxIndex)
    }
    let flow = flows.view[newIndex]
    return select(flow ? flow.id : undefined)
}


export function resume(flow) {
    return dispatch => fetchApi(`/flows/${flow.id}/resume`, { method: 'POST' })
}

export function resumeAll() {
    return dispatch => fetchApi('/flows/resume', { method: 'POST' })
}

export function kill(flow) {
    return dispatch => fetchApi(`/flows/${flow.id}/kill`, { method: 'POST' })
}

export function killAll() {
    return dispatch => fetchApi('/flows/kill', { method: 'POST' })
}


export function remove(flow) {
    return dispatch => fetchApi(`/flows/${flow.id}`, { method: 'DELETE' })
}

export function duplicate(flow) {
    return dispatch => fetchApi(`/flows/${flow.id}/duplicate`, { method: 'POST' })
}

export function replay(flow) {
    return dispatch => fetchApi(`/flows/${flow.id}/replay`, { method: 'POST' })
}

export function revert(flow) {
    return dispatch => fetchApi(`/flows/${flow.id}/revert`, { method: 'POST' })
}

export function update(flow, data) {
    return dispatch => fetchApi.put(`/flows/${flow.id}`, data)
}

export function uploadContent(flow, file, type) {
    const body = new FormData()
    file       = new window.Blob([file], { type: 'plain/text' })
    body.append('file', file)
    return dispatch => fetchApi(`/flows/${flow.id}/${type}/content.data`, { method: 'POST', body })
}


export function clear() {
    return dispatch => fetchApi('/clear', { method: 'POST' })
}

export function download() {
    window.location = '/flows/dump'
    return { type: REQUEST_ACTION }
}

export function upload(file) {
    const body = new FormData()
    body.append('file', file)
    return dispatch => fetchApi('/flows/dump', { method: 'POST', body })
}


export function select(id) {
    return {
        type: SELECT,
        flowIds: id ? [id] : []
    }
}
