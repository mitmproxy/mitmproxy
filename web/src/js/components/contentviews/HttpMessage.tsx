import React, {useCallback, useMemo, useRef, useState} from "react";
import {HTTPFlow, HTTPMessage} from "../../flow";
import {useAppDispatch, useAppSelector} from "../../ducks";
import {setContentViewFor} from "../../ducks/ui/flow";
import {ContentViewData, useContent} from "./useContent";
import {MessageUtils} from "../../flow/utils";
import FileChooser from "../common/FileChooser";
import * as flowActions from "../../ducks/flows";
import {uploadContent} from "../../ducks/flows";
import Button from "../common/Button";
import CodeEditor from "./CodeEditor";
import LineRenderer from "./LineRenderer";
import ViewSelector from "./ViewSelector";


type HttpMessageProps = {
    flow: HTTPFlow
    message: HTTPMessage
}

export default function HttpMessage({flow, message}: HttpMessageProps) {
    const dispatch = useAppDispatch();
    const part = flow.request === message ? "request" : "response";
    const contentView = useAppSelector(state => state.ui.flow.contentViewFor[flow.id + part] || "Auto");
    const editorRef = useRef<CodeEditor>(null);
    const [maxLines, setMaxLines] = useState<number>(useAppSelector(state => state.options.content_view_lines_cutoff));
    const showMore = useCallback(() => setMaxLines(Math.max(1024, maxLines * 2)), [maxLines]);
    const [edit, setEdit] = useState<boolean>(false);
    let url;
    if (edit) {
        url = MessageUtils.getContentURL(flow, message);
    } else {
        url = MessageUtils.getContentURL(flow, message, contentView, maxLines + 1);
    }
    const content = useContent(url, message.contentHash);
    const contentViewData = useMemo<ContentViewData | undefined>(() => {
        if (content && !edit) {
            try {
                return JSON.parse(content)
            } catch (e) {
                const err: ContentViewData = {"description": "Network Error", lines: [[["error", `${content}`]]]};
                return err;
            }
        } else {
            return undefined
        }
    }, [content]);

    if (edit) {
        const save = async () => {
            const content = editorRef.current?.getContent();
            await dispatch(flowActions.update(flow, {[part]: {content}}));
            setEdit(false);
        }
        return (
            <div className="contentview" key="edit">
                <div className="controls">
                    <h5>[Editing]</h5>
                    <Button onClick={save} icon="fa-check text-success" className="btn-xs">Done</Button>
                    &nbsp;
                    <Button onClick={() => setEdit(false)} icon="fa-times text-danger"
                            className="btn-xs">Cancel</Button>
                </div>
                <CodeEditor ref={editorRef} initialContent={content || ""}/>
            </div>
        );
    } else {
        let desc = contentViewData ? contentViewData.description : "Loading...";
        return (
            <div className="contentview" key="view">
                <div className="controls">
                    <h5>{desc}</h5>
                    <Button onClick={() => setEdit(true)} icon="fa-edit" className="btn-xs">Edit</Button>
                    &nbsp;
                    <FileChooser
                        icon="fa-upload"
                        text="Replace"
                        title="Upload a file to replace the content."
                        onOpenFile={content => dispatch(uploadContent(flow, content, part))}
                        className="btn btn-default btn-xs"/>
                    &nbsp;
                    <ViewSelector value={contentView}
                                  onChange={cv => dispatch(setContentViewFor(flow.id + part, cv))}/>
                </div>
                {ViewImage.matches(message) && <ViewImage flow={flow} message={message}/>}
                <LineRenderer lines={contentViewData?.lines || []}
                              maxLines={maxLines}
                              showMore={showMore}/>
            </div>
        );
    }
}


const isImage = /^image\/(png|jpe?g|gif|webp|vnc.microsoft.icon|x-icon)$/i
ViewImage.matches = msg => isImage.test(MessageUtils.getContentType(msg) || "")

type ViewImageProps = {
    flow: HTTPFlow,
    message: HTTPMessage,
}

export function ViewImage({flow, message}: ViewImageProps) {
    return (
        <div className="flowview-image">
            <img src={MessageUtils.getContentURL(flow, message)} alt="preview" className="img-thumbnail"/>
        </div>
    )
}
