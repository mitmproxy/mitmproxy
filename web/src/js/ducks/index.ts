import {
    AnyAction,
    applyMiddleware,
    combineReducers,
    compose,
    createStore as createReduxStore,
    PreloadedState,
    Store
} from "redux"
import eventLog from "./eventLog"
import flows from "./flows"
import ui from "./ui/index"
import connection from "./connection"
import options from './options'
import commandBar from "./commandBar";
import thunk, {ThunkAction, ThunkDispatch, ThunkMiddleware} from "redux-thunk";
import {TypedUseSelectorHook, useDispatch, useSelector} from "react-redux";
import backendState from "./backendState";
import options_meta from "./options_meta";


// @ts-ignore
const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;

export const rootReducer = combineReducers({
    commandBar,
    eventLog,
    flows,
    connection,
    ui,
    options,
    options_meta,
    backendState,
});
export type RootState = ReturnType<typeof rootReducer>


export type AppThunk<ReturnType = void> = ThunkAction<ReturnType,
    RootState,
    unknown,
    AnyAction>

export const createAppStore = (preloadedState?: PreloadedState<RootState>): Store<RootState> => {
    return createReduxStore(
        rootReducer,
        preloadedState,
        composeEnhancers(applyMiddleware(
            thunk as ThunkMiddleware<RootState>
        )))
};

export const store = createAppStore(undefined);

// this would be correct, but PyCharm bails on it
// export type AppDispatch = typeof store.dispatch
// instead:
export type AppDispatch = ThunkDispatch<RootState, void, AnyAction>
export const useAppDispatch = () => useDispatch<AppDispatch>()
export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector
