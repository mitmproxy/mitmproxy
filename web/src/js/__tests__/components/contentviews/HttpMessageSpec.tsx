import {TFlow} from "../../ducks/tutils";
import * as React from 'react';
import HttpMessage, {ViewImage} from '../../../components/contentviews/HttpMessage'
import {fireEvent, render, screen, waitFor} from "../../test-utils"
import fetchMock, {enableFetchMocks} from "jest-fetch-mock";

jest.mock("../../../contrib/CodeMirror")

enableFetchMocks();

test("HttpMessage", async () => {
    const lines = Array(512).fill([["text", "data"]]).concat(
        Array(512).fill([["text", "additional"]])
    );

    fetchMock.mockResponses(
        JSON.stringify({
            lines: lines.slice(0, 512 + 1),
            description: "Auto"
        }), JSON.stringify({
            lines,
            description: "Auto"
        }), JSON.stringify({
            lines: Array(5).fill([["text", "rawdata"]]),
            description: "Raw",
        }),
        "raw content",
        JSON.stringify({
            lines: Array(5).fill([["text", "rawdata"]]),
            description: "Raw",
        })
    );

    const tflow = TFlow();
    const {asFragment} = render(<HttpMessage flow={tflow} message={tflow.request}/>);
    await waitFor(() => screen.getAllByText("data"));
    expect(screen.queryByText('additional')).toBeNull();

    fireEvent.click(screen.getByText("Show more"));
    await waitFor(() => screen.getAllByText("additional"));

    fireEvent.click(screen.getByText("auto"));
    fireEvent.click(screen.getByText("raw"));
    await waitFor(() => screen.getAllByText("rawdata"));
    expect(asFragment()).toMatchSnapshot();

    fireEvent.click(screen.getByText("Edit"));
    expect(asFragment()).toMatchSnapshot();
    fireEvent.click(screen.getByText("Cancel"));

    await waitFor(() => screen.getAllByText("rawdata"));
    expect(asFragment()).toMatchSnapshot();
});

test("ViewImage", async () => {
    const flow = TFlow();
    const {asFragment} = render(<ViewImage flow={flow} message={flow.request}/>)
    expect(asFragment()).toMatchSnapshot();
});
