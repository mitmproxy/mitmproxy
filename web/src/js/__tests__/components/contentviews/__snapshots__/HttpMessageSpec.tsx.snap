// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HttpMessage 1`] = `
<DocumentFragment>
  <div
    class="contentview"
  >
    <div
      class="controls"
    >
      <h5>
        Raw
      </h5>
      <button
        class="btn-xs btn btn-default"
      >
        <i
          class="fa fa-clipboard"
        />
         Copy
      </button>
       
      <button
        class="btn-xs btn btn-default"
      >
        <i
          class="fa fa-edit"
        />
         Edit
      </button>
       
      <a
        class="btn btn-default btn-xs"
        href="#"
        title="Upload a file to replace the content."
      >
        <i
          class="fa fa-fw fa-upload"
        />
        Replace
        <input
          class="hidden"
          type="file"
        />
      </a>
       
      <a
        class="btn btn-default btn-xs"
        href="#"
      >
        <span>
          <i
            class="fa fa-fw fa-files-o"
          />
           
          <b>
            View:
          </b>
           raw 
          <span
            class="caret"
          />
        </span>
      </a>
    </div>
    <pre>
      <div>
        <span
          class="text"
        >
          rawdata
        </span>
      </div>
      <div>
        <span
          class="text"
        >
          rawdata
        </span>
      </div>
      <div>
        <span
          class="text"
        >
          rawdata
        </span>
      </div>
      <div>
        <span
          class="text"
        >
          rawdata
        </span>
      </div>
      <div>
        <span
          class="text"
        >
          rawdata
        </span>
      </div>
    </pre>
  </div>
</DocumentFragment>
`;

exports[`HttpMessage 2`] = `
<DocumentFragment>
  <div
    class="contentview"
  >
    <div
      class="controls"
    >
      <h5>
        [Editing]
      </h5>
      <button
        class="btn-xs btn btn-default"
      >
        <i
          class="fa fa-check text-success"
        />
         Done
      </button>
       
      <button
        class="btn-xs btn btn-default"
      >
        <i
          class="fa fa-times text-danger"
        />
         Cancel
      </button>
    </div>
    <div
      class="codeeditor"
    />
  </div>
</DocumentFragment>
`;

exports[`HttpMessage 3`] = `
<DocumentFragment>
  <div
    class="contentview"
  >
    <div
      class="controls"
    >
      <h5>
        Raw
      </h5>
      <button
        class="btn-xs btn btn-default"
      >
        <i
          class="fa fa-clipboard"
        />
         Copy
      </button>
       
      <button
        class="btn-xs btn btn-default"
      >
        <i
          class="fa fa-edit"
        />
         Edit
      </button>
       
      <a
        class="btn btn-default btn-xs"
        href="#"
        title="Upload a file to replace the content."
      >
        <i
          class="fa fa-fw fa-upload"
        />
        Replace
        <input
          class="hidden"
          type="file"
        />
      </a>
       
      <a
        class="btn btn-default btn-xs"
        href="#"
      >
        <span>
          <i
            class="fa fa-fw fa-files-o"
          />
           
          <b>
            View:
          </b>
           raw 
          <span
            class="caret"
          />
        </span>
      </a>
    </div>
    <pre>
      <div>
        <span
          class="text"
        >
          rawdata
        </span>
      </div>
      <div>
        <span
          class="text"
        >
          rawdata
        </span>
      </div>
      <div>
        <span
          class="text"
        >
          rawdata
        </span>
      </div>
      <div>
        <span
          class="text"
        >
          rawdata
        </span>
      </div>
      <div>
        <span
          class="text"
        >
          rawdata
        </span>
      </div>
    </pre>
  </div>
</DocumentFragment>
`;

exports[`HttpMessage 4`] = `
<DocumentFragment>
  <div
    class="contentview"
  >
    <div
      class="controls"
    >
      <h5>
        Raw
      </h5>
      <button
        class="btn-xs btn btn-default"
      >
        <i
          class="fa fa-clipboard"
        />
         Copy
      </button>
       
      <button
        class="btn-xs btn btn-default"
      >
        <i
          class="fa fa-edit"
        />
         Edit
      </button>
       
      <a
        class="btn btn-default btn-xs"
        href="#"
        title="Upload a file to replace the content."
      >
        <i
          class="fa fa-fw fa-upload"
        />
        Replace
        <input
          class="hidden"
          type="file"
        />
      </a>
       
      <a
        class="btn btn-default btn-xs"
        href="#"
      >
        <span>
          <i
            class="fa fa-fw fa-files-o"
          />
           
          <b>
            View:
          </b>
           raw 
          <span
            class="caret"
          />
        </span>
      </a>
    </div>
    <pre>
      <div>
        <span
          class="text"
        >
          rawdata
        </span>
      </div>
      <div>
        <span
          class="text"
        >
          rawdata
        </span>
      </div>
      <div>
        <span
          class="text"
        >
          rawdata
        </span>
      </div>
      <div>
        <span
          class="text"
        >
          rawdata
        </span>
      </div>
      <div>
        <span
          class="text"
        >
          rawdata
        </span>
      </div>
    </pre>
  </div>
</DocumentFragment>
`;

exports[`ViewImage 1`] = `
<DocumentFragment>
  <div
    class="flowview-image"
  >
    <img
      alt="preview"
      class="img-thumbnail"
      src="./flows/d91165be-ca1f-4612-88a9-c0f8696f3e29/request/content.data"
    />
  </div>
</DocumentFragment>
`;
